aquarium-native/src/opengl/FishModelGL.h

```diff
@@ -0,0 +1,78 @@
+// FishModelGL.h: Defines fish model of OpenGL.
+
+#pragma once
+#ifndef FishModelGL_H
+#define FishModelGL_H 1
+
+#include <string>
+
+#include "../FishModel.h"
+#include "ContextGL.h"
+#include "ProgramGL.h"
+
+class FishModelGL : public FishModel
+{
+  public:
+    FishModelGL(const ContextGL *context, Global *g, MODELGROUP type, MODELNAME name, bool blend);
+    void applyUniforms() const override;
+    void updatePerInstanceUnidorms() const override;
+    void updateFishCommonUniforms(float fishLength,
+                                  float fishBendAmount,
+                                  float fishWaveLength) override;
+    void init() override;
+    void applyTextures() const override;
+    void applyBuffers() const override;
+    void draw() const override;
+
+    void updateFishPerUniforms(float x,
+                               float y,
+                               float z,
+                               float nextX,
+                               float nextY,
+                               float nextZ,
+                               float scale,
+                               float time) override;
+
+    std::pair<float *, int> viewInverseUniform;

```

Comment:
In order to better match the JavaScript sources, don't you think it would be more of a fair comparison, and less verbose, to come up with some abstraction for uniforms so that there aren't std::pairs everywhere? The JavaScript sources use glGetActiveUniform and introspect on the type. How about doing something similar in this port?

Reply:
Hi Ken, sorry, I didn't explain clearly. we have a logically mapping version too, which is very similar to the webgl aquarium. It uses glGetActiveUniform and std::map to store the uniform values and its index, but we didn't upload the logically mapping version in this pull request.

This version has  optimized  the data structure and it's goal is to reach the best performance on desktop.  And we also want to put different backends into the same solution such as dawn and vulkan. We removed glGetActiveUniform in this one because dawn and vulkan doesn't support shader reflection. And maps are changed to std::pair because the uniforms will be replaced by uniform blocks in webgl2 version and other backends. So the data structure is changed to make opengl, dawn, vulkan, d3d12 coexit in the same solution.

 We can upload a logically mapping opengl backend as well but I'm not sure if it could coexit with dawn or vulkan since the data structure is very different, either can we put them in two solutions.

Reply:
Thanks @JiangYizhou for your explanation. At first, to have the most direct comparison between the current JavaScript WebGL Aquarium and native code, I think it would be best to upload your logical mapping version. We can add new versions later (maybe in other repositories?) which map better to Vulkan or Dawn. I'm most interested in exposing problems in WebGL's draw call validation performance, and the version which is closer in structure to the original JavaScript will be a better basis of comparison. Also, perhaps we'll find a way to evolve both the JavaScript and native versions for improved performance (for example, by using Uniform Buffer Objects in a WebGL 2.0 upgrade of the Aquarium).

Would you please revise this pull request to include your logical mapping version instead? Thank you.